### DISCUSSION TITLE
Bios import

### DISCUSSION THREAD

**Discussion by graschbaten** [retronas/discussions#59] [Q&A]
Is there a Bios import script?
I only have the latest complete Batocera Bios pack and complete Mister bioses.

How easly can i handle the RetroNAS Bios folder?

thx

**Comment by sairuk** [comment]
I don't know what those bios packs are, but you should just be able to copy them into the mister/batocera share providing their folder structure matches the project directories.

There is no tool for Bios import specifically at this stage, afaik no-one has ever written something we can incorporate, The project(s) should/could release a datfile or SMDB so their users could use existing tools like clrmame, rommanger, sabretools, etc to manage their bios then it'd be easy to add something at our end

if you come across one we can look at it, but it'd need to be managed, not just a oneshot,, 

To existing functionality
- Some of the smdb sets include bios sets though so that is "handled" as part of romimport
- MiSTer update_all used to have a bios getter so it RetroNAS was mounted it would download to RetroNAS
- SabreTools with a dat file (if one exists)

**Comment by codefaux** [comment]
Has there been movement on this? I've been trying to figure out how to get RetroNAS to "do" BIOS files for like an hour now.

Also in `/mnt/retronas/*/bios/` and `/mnt/retronas/bios/` there's a mess of bad symlinks? Literally every folder has either more folders or relative symlinks to...relative symlinks _pointing at themselves_?

FOR EXAMPLE,
`/mnt/retronas/emudeck/bios/nintendo/nes/famicom` is a link to `../../bios/nintendo/famicom` which is `/mnt/retronas/bios/nintendo/nes/famicom` but THAT is a symlink to `../../bios/nintendo/nes/famicom` -- aka ITSELF? Isn't that supposed to be a folder? `nes/famicom` also has a `disk` folder in it, and `inside the that folder` is a symlink named `disk` which points to `../../bios/nintendo/famicom/disk` (note we lost the 'nes' part of the path?) which is a relative symlink pointing to ... `../../bios/nintendo/famicom/disk` -- _itself again?_

So...why create this disaster if it doesn't support BIOS files in the first place? Can we correct it so that at least one time somewhere is written "You're on your own for BIOS organization"? I would also love for the system to no longer create folders it fails to use or acknowledge, let alone populate sensibly.

**Reply by sairuk** [reply]
the symlink recursion was a bug I'd thought we'd addressed previously however after review this morning it had not been applied to the cifs installer specific structure tasks ... a fix has been rolled out, we will also attempt to clean up bad links in our structure at install time now.

recursion issue above aside. the generic bios folders are a retronas structure, the resulting project (emudeck in your case ... mister, batocera etc for others) folders should be representative of the project requirements at point in time. 

if they are not that should be reported properly (disaster? lol) or a pull request created to address it.

We have no plans to write a bios import tool at this stage, community is still welcome to do so.

**Reply by codefaux** [reply]
This fix may be an issue on my system, as I rewrote all symlinks to be relative (so I can mount it via other filesystems) and also removed all broken symlinks myself, but I suppose that's actively my own fault in this case, lol. I'm prepared to reckon with it.

Good to know that it's at least a template layout -- may I suggest an info linked to the root of each target indicating "this is a template and must be self-populated with properly sourced BIOS files" or similar? Presenting the user with folder structure is best associated with an explanation of why it's there, and after all, the rest of the system is based around automatic operation.

> (disaster? lol)

Respectfully; my filesystem _was a disaster site._ I had to do triage and cleanup, and it was in every single folder, even the ROM folders. If it wasn't so much work I would've SO MUCH preferred to simply delete the whole thing. I'm offended that you'd mock my softballing. Your system is certainly not a disaster, but it created a big ugly mess in my filesystem which was perplexing and incredibly difficult to troubleshoot and repair from an outside, unfamiliar perspective.

I appreciate that BIOS import is outside the scope of this project, and I understand the reasoning behind it. I'm not demanding an import tool. I'm saying, document that in the same places you talk about importing ROMs so it's known. Ideally also a single non-platform-specific readme that says "DIY" in common language. Thank you for your attention.

**Reply by sairuk** [reply]
ahh ok, yes if you choose to move away from our standard generic folder structure, stuff will get nuked on reinstall. If we don't provide a suitable solution based on your desired structure this project may not be for you. For cleanup we would've have advised using `find <retronas path> -xtype l -exec rm {} \;`  instead of a manual clean up and wait for a fix had the issue been raised as such.

ultimately the name `roms` for that folder would be better as `software` these days from a retronas standpoint since folks also wanted bios, saves and savestate support added for MiSTer after the initial release. Implementing that then slowed down folder generation to well over 4min for (MiSTer) folders on an rpi4 so we tried a different approach to speed things up and inadvertently introduced the symlink recursion bug.

The only reason the we haven't added bios support to the retronas implementation of romimport is the project data that underpins the underlying project hasn't released dedicated per system bios SMDBs for use. If they did romimport could be expanded to support bios population. There is absolutely nothing stopping the community building these (or logiqx compartible dats) and putting them on github so they can be used with romimport or the other tools we provide installers for.

That said, we provide multiple tools that can/could automate population of folders, we don't intend to dictate how it can be done although we do automate the installation and setup,  some documentation and tooling for basic setup and operation. 
- samba/nfs/s3 etc for manual copy / scanning tools network endpoint
- romimport for software SmokeMonster dbs
- SabreTools for logiqx compatible dats (plus other format support)
- ucon64 which had some kind of scanning capability iirc but its been a few years since i used it

for "scanning tools network endpoint" that would be something like run cmpro and point the rebuilder/scanner at the samba share with bios dats for the systems you want.

all that said we can probably add a text file to the root of these folders at some point, can't hurt
